compose/service.py

```diff
@@ -619,3 +623,28 @@ def split_port(port):
 
     external_ip, external_port, internal_port = parts
     return internal_port, (external_ip, external_port or None)
+
+
+def build_extra_hosts(extra_hosts_config):
+    if not extra_hosts_config:
+        return {}
+
+    if isinstance(extra_hosts_config, list):
+        extra_hosts_dict = {}
+        for extra_hosts_line in extra_hosts_config:
+            if not isinstance(extra_hosts_line, six.string_types):
+                raise ConfigError(
+                    "extra_hosts_config \"%s\" must be either a list of strings or a string->string mapping," %
+                    extra_hosts_config
+                )
+            host, ip = extra_hosts_line.split(':')
+            extra_hosts_dict.update({host.strip(): ip.strip()})
+        extra_hosts_config = extra_hosts_dict
+
+    if isinstance(extra_hosts_config, dict):
+        return extra_hosts_config
```

Comment:
It might be a good idea to only support one of these formats.  We've supported both in the past, but I'm not sure that's necessary going forward. It seems to cause confusion and extra code.

If we decide to keep lists (or both), I would suggest `<host>=<ip>` (equals instead of colon) syntax instead. If you forget the quotes it confuses yaml into things its a mapping, and that can be very counter-intuitive to users.  If we use an equals, it should understand that it is a string, not a mapping.

If we decide to keep mappings (or both), we should probably do some basic validation of the keys and values, maybe even just assert that they are Truthy (not None or empty strings).